doc/man/pvfs2-remove-object.1
.TH PVFS2-REMOVE-OBJECT 1 2017-07-03
.SH NAME
\fBpvfs2-remove-object\fR \(en remove internal object from OrangeFS
volume
.SH SYNOPSIS
\fBpvfs2-remove-object\fR [\fB\-hv\fR] [\fB\-f \fIfs_id\fR]
[\fB\-o \fIhandle\fR] [\fB\-p \fIhandle \-d \fIentry_name\fR]
.SH DESCRIPTION
The
.B pvfs2-remove-object
removes an object from an OrangeFS volume by handle number.  This would
be used to repair a corrupt file system.  Improper use of this utility
could corrupt a file system.
.PP
The
.B -p
and
.B -d
options must be used together to remove a directory entry.
.PP
The options are as follows:
.IP -h
Display synopsis.
.IP -v
Print version number and exit.
.IP -f
Use
.I fs_id
as file system ID.
.IP -o
Remove the handle
.IR handle .
.IP -p
Specify the parent handle
.IR handle .
.IP -p
Specify the dirent name
.IR entry_name .
.SH ENVIRONMENT
.IP PVFS2_DEBUGFILE
If set to the path of a local file, redirect debug output to it.
.IP PVFS2_DEBUGMASK
Set the OrangeFS debug mask.  Possible masks are documented in
.BR pvfs2-set-debugmask ( 1 ) \& .
.IP PVFS2TAB_FILE
If set, the full pathname for an alternate
.IR pvfs2tab
file
.SH FILES
.I /etc/pvfs2tab
.SH EXAMPLES
Copy the file
.I /etc/motd
to
.IR /mnt/bar .
If
.I /mnt
is an OrangeFS volume, this will copy the file into OrangeFS.
.PP
.RS 6n
pvfs2-cp /etc/motd /mnt/bar
.RE
.SH CAVEATS
When
.I dst_file
is a directory the utility does not work.  In this respect it is
different than the
.BR cp ( 1 )
utility.
.SH BUGS
Please submit bug reports to devel@lists.orangefs.org
.SH SEE ALSO
.BR pvfs2-stat ( 1 ),
.BR pvfs2tab ( 5 )


ChangeLog
-----------------------

OrangeFS Release ChangeLog
-----------------------

orangefs-2.10.0 (Git Branch v.2.10.0)
https://github.com/waltligon/orangefs.git
up to commit-id 7c28362b

New functionality:

* - general updates
	1. numerous autotools related m4 files are updated.
	2. code refactored in numerous places to eliminate
	   compiler warnings from new versions of gcc.
	3. documentation updates.

* - Server
	1. updated formatting to conform to standard in simple-stripe code.

* - Clients
	1. build liborangefs by default. liborangefs provides an orangefs
	   version of posix system calls.

        2. Windows Client:
		removed security modes
	        added version resource to exe.
	        updated compiler options to output assembler listing.
	        updated installer
	        updated Event Log message files
	        add a service description when installing service.
	        changes to service startup and shutdown
	        switched messages to manifest file
	        improved config file processing.
	        retargeted test projects to Visual Studio 2022.
	        updated pvfs2-win-cp for Visual Studio 2022.
	        updated OpenSSL
		minor changes to GetVolumeInformation.
		updated GetFileInfo and CreateFile
		Improvements to error codes
		added get-file-info utility

* - bmi
	Handle IBV_EVENT_WQ_FATAL
	Handle IBV_EVENT_WQ_FATAL
	Add reference counting changes from bmi_rdma
	Build bmi_ib OR bmi_ib_exp, not both
	Add experimental BMI InfiniBand module
	Remove return_active_nic_handle()
	Refactor RDMA data structures and some functionality
	



* - general fixes
	Fix hostname message for RDMA protocol in pvfs2-genconfig
	Fix warning about using fast networks and TCP in pvfs2-genconfig
	Fix default RDMA port in pvfs2-genconfig usage message
	Add "ib_exp" protocol to pvfs2-genconfig options
	

* - Client fixes
        1. Windows Client:
		Fixed the failure to delete files in some cases.
	        fixed CreateFile by converting kernel values to usermode.
	        error code and directory-handling fixes.
	        fixed warnings in statecomp project.
		

* - bmi fixes
	Add server shutdown fix from bmi_rdma
	Add RDMA write refcnt fix from bmi_rdma
	Fix segfault issues:
             - when shutting down restarted servers
	     - Check if already disconnected in BMI_rdma_cancel()
	Add "options" parameter to BMI_rdma_initialize()
	Fix issue causing servers to hang on shutdown
	Only increment refcnt when an RDMA write is completely done

-----------------------
Development 2.9.9
orangefs-2.9.8 (Git Branch v.2.9.x)
https://github.com/waltligon/orangefs.git
up to rev 65ab0d2

This version of OrangeFS includes many bug fixes.

	1. Disable usrint by default.
	2. Fix potential memory overwrites in various utilities.
		- Client-core
		- Showcoll
		- Mkspace
		- Xattr
		- Viewdist
		- Str-utils
	3. Eliminate compiler warnings.
		- State machines
		- Trove
	4. Improve parsing of config file.
	5. Update documentation flow.
	6. Updated readmes.
	7. Migrate documentation to hugo format.
	8. Update perf mon example.



New functionality:
	1.  Added RoCE v1 support to our IB suite.


* - Server 
	1.  Fixed problems with crdirent and rename when using
	    distributed directories.
	2.  Fixed intermittent problems with updating timestamps.

* - Clients
	1.  pvfs2-stat.c
		When the object is a symlink, we now display the string length
		of the target in the size field, making the output consistent
		with the linux kernel.  Corrected call to sys-lookup so that
		symbolic links are followed correctly when -L is specified on
		the command line.
	2.  System Interface
		1.  Fixed problems with crdirent and rename when using distributed
		    directories.
		2.  Fixed intermittent problems with updating timestamps.

* - BMI
	1.  Part of adding RoCE v1 support was to make the IB port used for
	    communication configurable.

* - Software Infrastructure

* - Testing
	1.  Generic tests, #467 and #477, of the xfstest suite were removed from our
	    testing because we do not support the functionality.

	2.  Updated our testing to use version 20170929 of LTP.  Updated
	    the list of testcases to be used.

* - Kernel Module




orangefs-2.9.7 (Subversion Branch v.2.9.x)
http://dev.orangefs.org/svn/orangefs/branches/v.2.9.x
up to rev 13115 in trunk


New functionality:
	1.  Added support for Oracle Linux 7.3 with one caveat:  Cannot use
	    OrangeFS as the backing filesystem for a user-defined loop device.
	    Support will be forth coming.

	2.  Added support for large LMDB databases using DBMaxSize config
	    parameter.  

	3.  Added scripts in the examples/profile.d directory that can be placed in 
	    /etc/profile.d.  These scripts add OrangeFS locations to
	    LIBRARY_PATH, C_INCLUDE_PATH, AND LD_LIBRARY_PATH.

	4.  Added support for Fedora 26.  Certificate security is not yet
	    compatible with OpenSSL 1.1 but work is in progress.

	5.  Official OrangeFS rpm is in the Fedora distribution.  The rpm will
	    be in either Fedora 27 or Fedora 28.  The rpm will also be in
	    their EPEL repository, allowing other RHEL distributions to access
	    it.

	6.  Added --with-lmdb to the configure script, so you can specify the
	    location of your LMDB installation.  Otherwise, LMDB delivered
	    with OrangeFS will be used.  NOTE: Either way, to use LMDB instead
	    of Berkely DB, you must specify --with-db-backend=lmdb on the
	    configure line.

	7.  pvfs2-genconfig will ask you where you want to put the server config 
	    file, with the default being the <installation directory>/etc/orangefs.conf.
	    You can still put the server config filename directly on the command line.

	8.  ACL processing is now done as specified by POSIX.


* - Server  
	* - Enhanced permission checking and job scheduling for tree
	    operations.
	* - Removed memory leaks from the getattr operation and restructured
	    keyval buffer management.
	* - Corrected problem with mgmt-create-root-dir related to having a

src/apps/admin/pvfs2-remove-object.c
/*
 * (C) 2001 Clemson University and The University of Chicago
 *
 * See COPYING in top-level directory.
 */

#include <unistd.h>
#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <sys/time.h>
#include <time.h>
#include <stdlib.h>
#include <getopt.h>

#include "pvfs2.h"
#include "pvfs2-mgmt.h"
#include "str-utils.h"
#include "pvfs2-internal.h"

#ifndef PVFS2_VERSION
#define PVFS2_VERSION "Unknown"
#endif

typedef struct
{
    int remove_object_only;
    PVFS_handle object_handle;
    PVFS_handle parent_handle;
    PVFS_fs_id fs_id;
    char dirent_name[PATH_MAX];
}  options_t;

static void usage(int argc, char **argv)
{
    fprintf(stderr, "Usage: %s [OPTION] ...\n\n", argv[0]);
    fprintf(stderr, "-h, --help                  display this help "
            "information and exit\n");
    fprintf(stderr, "-v, --version               display version "
            "information and exit\n");
    fprintf(stderr, "-f, --fsid=<fs_id>          use the specified "
            "fs_id for the specified handle\n");
    fprintf(stderr, "-o, --object=<handle>       remove the specified "
            "object handle\n");
    fprintf(stderr, "-p, --parent=<handle>       use the specified parent "
            "handle\n\t\t\t\t(requires -d, or --dirent option)\n");
    fprintf(stderr, "-d, --dirent=<entry_name>   use the specified dirent "
            "name\n\t\t\t\t(requires -p, or --parent option)\n");
    fprintf(stderr, "\nNOTE:  The -f, or --fsid option must ALWAYS "
            "be specified\n");
}

static options_t *parse_args(int argc, char **argv)
{
    int ret = 0, option_index = 0;
    const char *cur_option = NULL;
    options_t *tmp_opts = NULL;
    static struct option long_opts[] =
    {
        {"help",0,0,0},
        {"version",0,0,0},
        {"object",1,0,0},
        {"parent",1,0,0},
        {"dirent",1,0,0},
        {"fsid",1,0,0},
        {0,0,0,0}
    };

    if (argc == 1)
    {
        return NULL;
    }

    tmp_opts = (options_t *)malloc(sizeof(options_t));
    if (!tmp_opts)
    {
	return NULL;
    }
    memset(tmp_opts, 0, sizeof(options_t));

    while((ret = getopt_long(argc, argv, "hvo:p:d:f:",
                             long_opts, &option_index)) != -1)
    {
	switch(ret)
        {
            case 0:
                cur_option = long_opts[option_index].name;

                if (strcmp("help", cur_option) == 0)
                {
                    goto do_help;
                }
                else if (strcmp("version", cur_option) == 0)
                {
                    goto do_version;
                }
                else if (strcmp("object", cur_option) == 0)
                {
                    goto do_object;
                }
                else if (strcmp("parent", cur_option) == 0)
                {
                    goto do_parent;
                }
                else if (strcmp("dirent", cur_option) == 0)
                {
                    goto do_dirent;
                }
                else if (strcmp("fsid", cur_option) == 0)
                {
                    goto do_fsid;
                }
                else
                {
                    usage(argc, argv);
                    exit(1);
                }
                break;
            case 'h':
          do_help:
                usage(argc, argv);
                exit(0);
                break;
	    case 'v':
          do_version:
                printf("%s\n", PVFS2_VERSION);
                exit(0);
		break;
	    case 'o':
          do_object:
                tmp_opts->remove_object_only = 1;
                tmp_opts->object_handle =
                    (PVFS_handle)strtoull(optarg, NULL,10);
		break;
	    case 'p':
          do_parent:
                tmp_opts->parent_handle =
                    (PVFS_handle)strtoull(optarg, NULL,10);
		break;
            case 'd':
          do_dirent:
                snprintf(tmp_opts->dirent_name, PATH_MAX, "%s", optarg);
                break;
	    case 'f':
          do_fsid:
                tmp_opts->fs_id =
                    (PVFS_fs_id)strtoull(optarg, NULL,10);
		break;
	    case '?':
		usage(argc, argv);
		exit(1);
	}
    }
    return tmp_opts;
}

int main(int argc, char **argv)
{
    int ret = -1;
    options_t *user_opts = NULL;
    PVFS_object_ref ref;
    PVFS_credential credentials;

    user_opts = parse_args(argc, argv);
    if (!user_opts)
    {
        usage(argc, argv);
	return ret;
    }

    ref.fs_id = user_opts->fs_id;
    if (user_opts->remove_object_only)
    {
        ref.handle = user_opts->object_handle;
        if ((ref.handle == PVFS_HANDLE_NULL) ||
            (ref.fs_id == PVFS_FS_ID_NULL))
        {
            fprintf(stderr, "Invalid object reference specified: "
                    "%llu,%d\n", llu(ref.handle), ref.fs_id);
            return ret;
        }
    }
    else
    {
        ref.handle = user_opts->parent_handle;
        if ((ref.handle == PVFS_HANDLE_NULL) ||
            (ref.fs_id == PVFS_FS_ID_NULL))
        {
            fprintf(stderr, "Invalid parent reference specified: "
                    "%llu,%d\n", llu(ref.handle), ref.fs_id);
            return ret;
        }

        /*
         * Space for struct user_opts points to was malloced and memset to
         * zero in parse_args.
         */
        if (strlen(user_opts->dirent_name) == 0)
        {
            fprintf(stderr, "No dirent name specified under parent "
                    "%llu,%d\n", llu(ref.handle), ref.fs_id);
            return ret;
        }
    }

    ret = PVFS_util_init_defaults();
    if (ret < 0)
    {
	PVFS_perror("PVFS_util_init_defaults", ret);
	return -1;
    }

    ret = PVFS_util_gen_credential_defaults(&credentials);
    if (ret < 0)
    {
        PVFS_perror("PVFS_util_gen_credential_defaults", ret);
        return -1;
    }

    if (user_opts->remove_object_only)
    {
        fprintf(stderr,"Attempting to remove object %llu,%d\n",
                llu(ref.handle), ref.fs_id);

        ret = PVFS_mgmt_remove_object(ref, &credentials, NULL);
        if (ret)
        {
            PVFS_perror("PVFS_mgmt_remove_object", ret);
        }
    }
    else
    {
        fprintf(stderr,"Attempting to remove dirent \"%s\" under %llu,%d"
                "\n", user_opts->dirent_name, llu(ref.handle), ref.fs_id);

        ret = PVFS_mgmt_remove_dirent(
            ref, user_opts->dirent_name, &credentials, NULL);
        if (ret)
        {
            PVFS_perror("PVFS_mgmt_remove_dirent", ret);
        }
    }

    free(user_opts);
    return ret;
}

/*
 * Local variables:
 *  c-indent-level: 4
 *  c-basic-offset: 4
 * End:
 *
 * vim: ts=8 sts=4 sw=4 expandtab
 */
